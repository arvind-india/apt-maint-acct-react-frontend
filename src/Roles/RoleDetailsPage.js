import React from 'react'
import { Redirect } from 'react-router-dom'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import Select from 'react-select';

import {
    Form,
    Button,
    FormGroup,
    FormText,
    Input,
    Label,
    Col,
    FormFeedback
} from 'reactstrap'

import { roleActions as actions, alertActions } from '../_actions'


class RoleDetailsPage extends React.Component {

  constructor(props) {
    super(props)
    const { dispatch, roleDetails } = props
    console.log('RoleDetails in constructor: ', roleDetails)
    this.state = {
      mModel: {}, // model being modified
      selectedOption: '',
      submitted: false,
      touched: false
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleMultiSelectChange = this.handleMultiSelectChange.bind(this)

    dispatch(actions.getAll())
    dispatch(alertActions.clear())  // clear alert messages from other pages
  }
  componentDidMount() {
    this.props.dispatch(actions.getById(this.props.match.params.id))
    //this.props.dispatch(actions.getAll())
  }
  handleSubmit(event) {
    event.preventDefault()
    this.setState({ submitted: true })
  }
  handleChange(event) {
    const { name, value } = event.target
    const { mModel } = this.state
    this.setState({
      [name]: value
    })
  }

  render() {
    const { roleDetails, user, match, alert, submitted } = this.props
    console.log('Role Details render props: ', this.props)
    let model = roleDetails
    return (
      <div>
        <h2>Role Details</h2>
        {alert.message && <div className={`alert ${alert.type}`}>{alert.message}</div>}
        {model.loading && <em>Loading user details...}</em>}
        {model.error && <span className="text-danger">{model.error}</span>}
        {model.data && this.show(model.data)}
      </div>
    )
  }
  show(data){

    return <Form onSubmit={this.handleSubmit} className="grid-form">
      <fieldset>
  			<legend>View or Edit</legend>
        <div data-row-span="2">
          {this.showRolename(data)}
          {this.showInherits(data)}
        </div>
        <div data-row-span="1">
          {this.showMultiSelect(data)}
        </div>
        <div data-row-span="1">
          {this.showDescription(data)}
        </div>
      </fieldset>
      <br/>
      <Button type="submit" color="primary">Save</Button>
      <Button color="link"><Link to="/roles">Cancel</Link></Button>
    </Form>
  }

  showRolename(data) {
    const { submitted, mModel } = this.state
    return <div data-field-span="1">
				<Label>Role Name</Label>
        <Input
          type="text"
          name="name"
          placeholder="Role name here"
          className="inputField"
          defaultValue={data.name}
          onChange={this.handleChange}
        />
        {submitted && mModel.name != null && mModel.name == ""
        && <FormText color="danger">Role-Name is required</FormText>}
			</div>
  }

  showInherits(data) {
    const { submitted, mModel } = this.state
    return <div data-field-span="1">
				<Label>Inherits</Label>
        <Input
          type="text"
          name="inherits"
          placeholder="Inherits here"
          className="inputField"
          defaultValue={data.inherits}
          onChange={this.handleChange}
        />
        {submitted && mModel.inherits != null && mModel.inherits == ""
        && <FormText color="danger">Role-Inherits is required</FormText>}
			</div>
  }

  showDescription(data) {
    const { submitted, mModel } = this.state
    return <div data-field-span="1">
				<Label>Description</Label>
        <Input
          type="text"
          name="description"
          placeholder="Description here"
          className="inputField"
          defaultValue={data.description}
          onChange={this.handleChange}
        />
        {submitted && mModel.description != null && mModel.description == ""
        && <FormText color="danger">Role-Description is required</FormText>}
			</div>
  }
  handleMultiSelectChange(selectedOption) {
    console.log('Selected Option: ', selectedOption)
    this.setState({ selectedOption })
  }
  showMultiSelect(data) {
    const { selectedOption } = this.state;
    const { roles } = this.props
/*
    console.log('Roles: ', roles.items)
    if(selectedOption == null) {
      this.setState({ selectedOption: data.inherits })
    }
*/
  	//const value = selectedOption && selectedOption.value;
    //const value = selectedOption
    return (
      <Select
        name="form-field-name"
        value={data.inherits}
        multi={true}
        joinValues={true}
        simpleValue={true}
        placeholder="Select Inherits..."
        onChange={this.handleMultiSelectChange}
        defaultValue="guest"
        valueKey="name"
        labelKey="name"
        options={roles.items}
      />
    );
  }
}

function mapStateToProps(state) {
  const { roles, roleDetails, authentication, alert } = state
  const { user } = authentication
  return {
    user,
    roles,
    roleDetails,
    alert
  }
}

const connectedDetailsPage = connect(mapStateToProps)(RoleDetailsPage)
export { connectedDetailsPage as RoleDetailsPage }
