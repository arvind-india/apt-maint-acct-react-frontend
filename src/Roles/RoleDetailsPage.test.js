// RolesPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { RoleDetails } from './RoleDetailsPage'

const mockRoles = [
  {
    id: 1,
    name: 'admin',
    description:'administrator',
    inherits: ''
  },
  {
    id: 2,
    name: 'owner',
    description:'owner of the flat',
    inherits: 'tenant'
  }

]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
//    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    roles: {
      items: []
    }
  },
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
//    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    roles: {
      items: []
    }
  }
]
const mockLocations = [
  {
    state: {
      model: mockRoles[0],
      submitted: false,
      adding: false
    }
  },
  {
    state: {
      model: mockRoles[1],
      submitted: false,
      adding: false
    }
  }
]

const mocks = [
  {
    name: {name: 'name', value: mockRoles[0].name},
    description: {name: 'description', value: mockRoles[0].description},
    inherits: mockRoles[0].inherits
  },
  {
    name: {name: 'name', value: mockRoles[1].name},
    description: {name: 'description', value: mockRoles[1].description},
    inherits: mockRoles[1].inherits
  }
]
const component = shallow(<RoleDetails {...mockProps[0]} location={mockLocations[0]}/>)

describe('RolesPage params test', () => {
  it('should render RolesPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('model')).toBeDefined()
  })
  it('should have submitted state', () => {
    expect(component.state('submitted')).toBeDefined()
  })
  it('should have adding state', () => {
    expect(component.state('adding')).toBeDefined()
  })
})

describe('RoleDetailsPage events test 1', () => {
/*  it('should call clearAlert on submit and no crash', () => {
    component.find('#roleDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[0].clearAlert.mock.calls.length).toEqual(1)
  }) */
  it('should be called with required states as arguments', () => {
    component.find('#name').simulate('change', {target: mocks[0].name})
    component.find('#description').simulate('change', {target: mocks[0].description})
    component.find('Select').simulate('change', mocks[0].inherits)
    component.find('#roleDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[0].saveChanges.mock.calls[0][0]).toEqual(mockRoles[0])
  })
})

const component2 = shallow(<RoleDetails {...mockProps[1]} location={mockLocations[1]}/>)
describe('RoleDetailsPage events test 2', () => {
  it('should be called with required states as arguments', () => {
    component2.find('#name').simulate('change', {target: mocks[1].name})
    component2.find('#description').simulate('change', {target: mocks[1].description})
    component2.find('Select').simulate('change', mocks[1].inherits)
    component2.find('#roleDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[1].saveChanges.mock.calls[0][0]).toEqual(mockRoles[1])
  })
})
