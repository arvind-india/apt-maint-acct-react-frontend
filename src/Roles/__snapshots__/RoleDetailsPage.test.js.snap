// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RolesPage params test should render initial layout 1`] = `
Array [
  <div>
    <h2>
        Role Details
    </h2>
    <div
        className="alert undefined"
    >
        testingAlert
    </div>
    <Form
        className="grid-form"
        id="roleDetailsForm"
        onSubmit={[Function]}
        tag="form"
    >
        <fieldset>
            <legend>
                Edit
            </legend>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Role Name
                    </Label>
                    <Input
                        className="inputField"
                        id="name"
                        name="name"
                        onChange={[Function]}
                        placeholder="Role name here"
                        tag="p"
                        type="text"
                        value="admin"
                    />
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Inherits (optional)
                    </Label>
                    <Select
                        arrowRenderer={[Function]}
                        autosize={true}
                        backspaceRemoves={true}
                        backspaceToRemoveMessage="Press backspace to remove {label}"
                        clearAllText="Clear all"
                        clearRenderer={[Function]}
                        clearValueText="Clear value"
                        clearable={true}
                        closeOnSelect={true}
                        deleteRemoves={true}
                        delimiter=","
                        disabled={false}
                        escapeClearsValue={true}
                        filterOptions={[Function]}
                        id="inherits"
                        ignoreAccents={true}
                        ignoreCase={true}
                        inputProps={Object {}}
                        isLoading={false}
                        joinValues={true}
                        labelKey="name"
                        matchPos="any"
                        matchProp="any"
                        menuBuffer={0}
                        menuRenderer={[Function]}
                        multi={true}
                        name="form-field-name"
                        noResultsText="No results found"
                        onBlurResetsInput={true}
                        onChange={[Function]}
                        onCloseResetsInput={true}
                        onSelectResetsInput={true}
                        openOnClick={true}
                        optionComponent={[Function]}
                        options={Array []}
                        pageSize={5}
                        placeholder="Select Inherits..."
                        removeSelected={true}
                        required={false}
                        rtl={false}
                        scrollMenuIntoView={true}
                        searchable={true}
                        simpleValue={true}
                        tabSelectsValue={true}
                        trimFilter={true}
                        value=""
                        valueComponent={[Function]}
                        valueKey="name"
                    />
                </div>
            </div>
            <div
                data-row-span="1"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Description
                    </Label>
                    <Input
                        className="inputField"
                        id="description"
                        name="description"
                        onChange={[Function]}
                        placeholder="Description here"
                        tag="p"
                        type="text"
                        value="administrator"
                    />
                </div>
            </div>
        </fieldset>
        <br />
        <Button
            color="primary"
            disabled={true}
            hidden={false}
            tag="button"
            title="Save changes"
            type="submit"
        >
            Save
        </Button>
        <Button
            color="link"
            tag="button"
        >
            <Link
                className="text-danger"
                replace={false}
                title="Go to Roles"
                to="/roles"
            >
                Cancel
            </Link>
        </Button>
    </Form>
</div>,
]
`;
